ssl/ssl_sess.c

```diff
@@ -1128,11 +1158,41 @@ static void SSL_SESSION_list_add(SSL_CTX *ctx, SSL_SESSION *s)
         s->prev = (SSL_SESSION *)&(ctx->session_cache_head);
         s->next = (SSL_SESSION *)&(ctx->session_cache_tail);
     } else {
-        s->next = ctx->session_cache_head;
-        s->next->prev = s;
-        s->prev = (SSL_SESSION *)&(ctx->session_cache_head);
-        ctx->session_cache_head = s;
+        t = SESS_TIMEOUT(s);
+        if (t >= SESS_TIMEOUT(ctx->session_cache_head)) {
```

Comment:
`difftime(3)` here too?

Reply:
The compare should be safe, it's the addition inside `SESS_TIMEOUT()` that's the problem.
`SESS_TIMEOUT()` adds a `time_t` plus a timeout value. This macro could be modified to provide a safe value in the case of overflow (I haven't look for the MAX constant). Alternatives might be to write a safe `time_t` + timeout function that handles all overflow issues without reverting to `double`s.

(TBH, I plan to retire before 2038...)